### ISSUE TITLE
Build Android Applications

### ISSUE THREAD

**Issue by non-Jedi** [NHDaly/ApplicationBuilder.jl#54]
I have only the foggiest idea of the prerequisites here, but I thought I'd throw the idea out there as a potential long-term goal.

[There's a thread on discourse on running Julia itself on Android](https://discourse.julialang.org/t/julia-on-ios-and-android/7346). The Android half of this is probably something to do with [Android NDK](https://developer.android.com/ndk/). Beyond that, we'd probably need some sort of GUI framework specifically designed with Android in mind (possibly based on [JavaCall.jl](https://github.com/JuliaInterop/JavaCall.jl)?). Here's a guide from Mozilla on building/deploying a Rust (Julia brother in LLVM) library on Android: https://mozilla.github.io/firefox-browser-architecture/experiments/2017-09-21-rust-on-android.html

I was thinking that IOS might be an easier initial target than Android since I believe the Swift toolchain for IOS uses LLVM anyway, but in my brief digging I didn't find much efforts in that direction. There is this repo where Rust is built as a *library* for both Android and IOS though: https://github.com/kennytm/rust-ios-android.

Basically, I got curious about the viability of ever running Julia apps on mobile platforms and your repository seemed like the most relevant place to brain dump a bunch of relevant links. Sorry. Feel free to close.

**Comment by NHDaly** [comment]
> I was thinking that IOS might be an easier initial target than Android since I believe the Swift toolchain for IOS uses LLVM anyway, but in my brief digging I didn't find much efforts in that direction. There is this repo where Rust is built as a library for both Android and IOS though: https://github.com/kennytm/rust-ios-android.

I think that it will be a _long time_ before we can build Julia for iOS. iOS has a strict requirement that you aren't allowed to compile code inside an app, and runtime compilation is sort of... julia's whole thing.
To be allowed in an iOS app, a julia program would need to be _fully precompiled_ for every type, with sufficient fallbacks for any surprising types, as it would not be allowed to do further specialization. I think we're still a ways off from that working well.

---------------

Running on Android doesn't seem too hard to me! It's just another linux, afterall. I just don't think there's been much focus on. I'd imagine a sustained effort for a couple of weeks would yield positive results! :)

Thanks for all those links!

I am indeed going to close this, but I'm glad this is recorded! Let us know if you (or anyone) does indeed try anything! :) Thanks!
